File: dev_ws/src/learning_action_cpp/src/fibonacci_client.cpp
/*
 * @Description: 
 * @version: v1.0
 * @Author: HTY
 * @Date: 2022-07-23 17:21:02
 */

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "rclcpp_components/register_node_macro.hpp"

#include "learning_interface/action/fibonacci.hpp"


namespace learning_action_cpp
{

class FibonacciClient : public rclcpp::Node
{
public:
    using Fibonacci = learning_interface::action::Fibonacci;
    using GoalHandleFibonacci = rclcpp_action::ClientGoalHandle<Fibonacci>;

    explicit FibonacciClient(const rclcpp::NodeOptions & options) : Node("fibonacci_client_node", options)
    {
        this->action_client = rclcpp_action::create_client<Fibonacci>(this, "fibonacci");
        this->timer = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&FibonacciClient::send_goal, this));
    }

    void send_goal()
    {
        using namespace std::placeholders;

        this->timer->cancel();

        if (!this->action_client->wait_for_action_server())
        {
            RCLCPP_ERROR(this->get_logger(), "Action server is not available");
            rclcpp::shutdown();
        }

        auto goal_msg = Fibonacci::Goal();
        goal_msg.order = 10;

        RCLCPP_INFO(this->get_logger(), "Sending goal");

        auto send_goal_options = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
        send_goal_options.goal_response_callback = std::bind(&FibonacciClient::goal_response_callback, this, _1);
        send_goal_options.feedback_callback = std::bind(&FibonacciClient::feedback_callback, this, _1, _2);
        send_goal_options.result_callback = std::bind(&FibonacciClient::result_callback, this, _1);
        this->action_client->async_send_goal(goal_msg, send_goal_options);
    }

private:
    rclcpp_action::Client<Fibonacci>::SharedPtr action_client;
    rclcpp::TimerBase::SharedPtr timer;

    void goal_response_callback(const GoalHandleFibonacci::SharedPtr & goal_handle)
    {
        if (!goal_handle)
        {
            RCLCPP_ERROR(this->get_logger(), "Goal rejected");
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "Goal accepted");
        }
    }

    void feedback_callback(const GoalHandleFibonacci::SharedPtr&, const std::shared_ptr<const Fibonacci::Feedback>& feedback)
    {
        std::stringstream ss;
        ss << "Next number in sequence received: ";
        for (auto number : feedback->partial_sequence)
        {
            ss << number << " ";
        }
        RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());
    }

    void result_callback(const GoalHandleFibonacci::WrappedResult & result)
    {
        switch (result.code)
        {
            case rclcpp_action::ResultCode::SUCCEEDED:
                break;
            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_ERROR(this->get_logger(), "Goal was cancelled");
            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
            default:
                RCLCPP_ERROR(this->get_logger(), "Unknown result code");
                return;
        }
        std::stringstream ss;
        ss << "Result received: ";
        for (auto number : result.result->sequence)
        {
            ss << number << " ";
        }
        RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());
        rclcpp::shutdown();
    }
};

}

RCLCPP_COMPONENTS_REGISTER_NODE(learning_action_cpp::FibonacciClient)

File: dev_ws/src/learning_action_python/learning_action_python/fibonacci_client.py
"""
 @Description: 
 @version: v1.0
 @Author: HTY
 @Date: 2022-07-23 16:20:53
"""

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from learning_interface.action import Fibonacci


class FibonacciClient(Node):
    def __init__(self, node):
        super(FibonacciClient, self).__init__(node)
        self.action_client = ActionClient(self, Fibonacci, "fibonacci")
        self.send_future = None
        self.result_future = None

    def send_goal(self, order):
        goal_msg = Fibonacci.Goal()
        goal_msg.order = order
        while not self.action_client.wait_for_server(timeout_sec=1.0):
            self.get_logger().info("Waiting for server started...")
        self.send_future = self.action_client.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        self.send_future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info("Goal rejected!")
            return

        self.get_logger().info("Goal accepted!")
        self.result_future = goal_handle.get_result_async()
        self.result_future.add_done_callback(self.get_result_callback)

    def get_result_callback(self, future):
        result = future.result().result
        self.get_logger().info(f"Result: {result.sequence}")

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info(f"Received feedback: {feedback.partial_sequence}")


def main(args=None):
    rclpy.init(args=args)
    action_client = FibonacciClient("fibonacci_client_node")
    action_client.send_goal(10)
    rclpy.spin(action_client)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
